python_modules/dagster/dagster/_core/definitions/asset_checks.py

```diff
@@ -127,3 +143,70 @@ def get_attributes_dict(self) -> Dict[str, Any]:
             specs=self._specs,
             input_output_props=self._input_output_props,
         )
+
+
+@experimental
+def build_blocking_asset_check(
+    asset_def: "AssetsDefinition",
+    checks: Sequence[AssetChecksDefinition],
+) -> "AssetsDefinition":
+    from dagster import AssetIn, In, OpExecutionContext, Output, op
+    from dagster._core.definitions.decorators.asset_decorator import graph_asset_no_defaults
+    from dagster._core.storage.asset_check_execution_record import AssetCheckExecutionRecordStatus
+
+    check_specs = []
+    for c in checks:
+        check_specs.extend(c.specs)
+
+    check_output_names = [c.get_python_identifier() for c in check_specs]
+
+    check.invariant(len(asset_def.op.output_defs) == 1)
+    asset_out_type = asset_def.op.output_defs[0].dagster_type
+
+    @op(ins={"materialization": In(asset_out_type), "check_evaluations": In(Nothing)})
+    def fan_in_checks_and_materialization(context: OpExecutionContext, materialization):
+        yield Output(materialization)
```

Comment:
This is also very worthy of explanation as it is quite confusing at first blush.

You may want to consider--for no other reason that clarifying things for code readers–-a couple alternative approches:

1) A wrapper class around `AssetMaterialization` that expresses the intent of the author––i.e. you––that you are shuffling this information between ops on purpose. Without more context, this looks like the author misunderstand concepts in the framework and screwed thing up
2) Since we ended up just fetching check evaluations from the instance anyways, just do that for the materialization as well.

I bias towards figuring out a way to make 1) work for all the events you are shuffling around via wrapper classes. However I don't have full contect on the asset check eval decisions. If there is a very good reason to do that, then let's make the materialization treatment similar.

Reply:
added a comment. I think the crux is getting the correct io manager output written, not passing the events around